src/progress/ProgressState.ts

```diff
@@ -4,31 +4,83 @@
  * We improved the Progress Component [Progress](https://github.com/chakra-ui/chakra-ui/tree/develop/packages/progress)
  * to work with Reakit System
  */
-import { isUndefined, valueToPercent } from "@chakra-ui/utils";
+import * as React from "react";
+import { SealedInitialState, useSealedState } from "reakit-utils";
 
-export interface UseProgressProps {
+import { valueToPercent, isFunction } from "../utils";
+
+type AriaValueText = string | ((value: number, percent: number) => string);
+
+export interface ProgressState {
   /**
    * The `value` of the progress indicator.
    * If `undefined` the progress bar will be in `indeterminate` state
+   * @default 0
    */
-  value?: number;
+  value: number;
   /**
    * The minimum value of the progress
+   * @default 0
+   */
+  min: number;
+  /**
+   * The maximum value of the
+   * @default 100
    */
-  min?: number;
+  max: number;
   /**
-   * The maximum value of the progress
+   * Set isInterminate state
+   * @default false
    */
-  max?: number;
+  isIndeterminate?: boolean;
+  /**
+   * Defines the human readable text alternative of aria-valuenow for a range widget.
+   */
+  ariaValueText: AriaValueText;
+  /**
+   * Percentage of the value progressed with respect to min & max
+   */
+  percent: number;
+}
+
+export interface ProgressAction {
+  /**
+   * Update the value of the progress indicator
+   */
+  setValue?: React.Dispatch<React.SetStateAction<number>>;
 }
 
-export const useProgressState = (props: UseProgressProps = {}) => {
-  const { value, min = 0, max = 100 } = props;
+export type ProgressInitialState = Partial<
+  Pick<
+    ProgressState,
+    "value" | "min" | "max" | "isIndeterminate" | "ariaValueText"
+  >
+>;
 
-  const percent = value != null ? valueToPercent(value, min, max) : undefined;
-  const isIndeterminate = isUndefined(value);
+export type ProgressStateReturn = ProgressState & ProgressAction;
```

Comment:
`export type ProgressStateReturn = ReturnType<typeof useProgressState>;` would be nicer i guess

Reply:
Well yeah we would not get the jsdocs in return state maybe we can keep it this way... i'll do some research to see if tsdoc supports any shortcuts for this with ReturnType generic

Reply:
I am thinking this is nice and better, both does the same.
Helps with clear separation between the state & Actions

Reply:
Too much manual types management. ReturnType is easier to manage since it would update depending on the hook's return values.

Reply:
I am just prefering it the Reakit's way. Yeah, Typescript will infer from the returns.
Since we already have the State & Actions defined, I prefered this useful.

This was helpful, when I am improving a featurn where I return more items, Typescript warns me to include those state/Action, so that I can respective JSDoc.

Reply:
Okay agreed we can keep it. Just that did not liked the idea of deriving both the `initial props` and `return state` from the same types, for this small example with progress it is fine, but it would be a little bit messy if we wanted to change/rename return values suppose we want to change the "setValue" to "setProgressValue" then we have to overwrite the Return types anyways.


It is fine for this progress component but not very effective for other components like Segment, or Calendar. where props and state are very different.

Reply:
Then we can define the `setProgressValue` in Action and it will be in the Return type.
Even I find `export type ProgressStateReturn = ReturnType<typeof useProgressState>;` this was better initially.

But on deep diving into the motivation of Reakit's approach, I find the latter approach useful. Even for ComboBox reakit has done the same.

Reply:
Ok we can keep it no change required